### DISCUSSION TITLE
Linter diagnostics

### DISCUSSION THREAD

**Discussion by bojohnson5** [scikit-hep/uproot5#1129] [Q&A]
I'm using ruff for linting my code and see these errors pop up on various lines. I'm curious to know if others have encountered this and if there's a way to fix it. Thanks!

![Screenshot 2024-02-16 at 3 56 41 PM](https://github.com/scikit-hep/uproot5/assets/20647190/86d4b5da-e5c8-4274-b10e-57f59994d8df)

**Comment by jpivarski** [comment]
If `cenns` is a ROOT directory, then the data type of `cenns['en']` etc. can't be known without reading the file. ruff does static linting, deducing the types of each variable by looking at the code and not running it—it can't possibly know if the data type of `cenns['en']` is a TTree (a type that has a `array` method).

Rather than admitting ignorance, ruff seems to be thinking that the return value of any `cenns.__getitem__` is also a directory (`ReadOnlyDirectory`). I don't know why it would think that: the `ReadOnlyDirectory.__getitem__` method doesn't even claim what its output type might be (because it can't):

https://github.com/scikit-hep/uproot5/blob/ea207f992c61bd2265e112f29c8b0c3d0cd6706d/src/uproot/reading.py#L2040

In general, we should be telling linters that we don't have the information they need. It seems like that's the case here—I don't know why ruff thinks it knows this type.

**Reply by bojohnson5** [reply]
Thanks for the reply. I'll try and dig a little more but for the meantime have turned off the ruff linting.

**Comment by jpivarski** [comment]
I'd like to keep all Discussions open, so that other people can find them. Thanks!
